File: apps/residence/lib/core/errors/app_error.dart
/**
 * Custom error classes for Residence app
 */

class AppError implements Exception {
  final String message;
  final String code;
  final dynamic details;

  AppError(this.message, {this.code = 'UNKNOWN_ERROR', this.details});

  @override
  String toString() => 'AppError: $message (code: $code)';
}

class AuthError extends AppError {
  AuthError(String message, {dynamic details})
      : super(message, code: 'AUTH_ERROR', details: details);
}

class NetworkError extends AppError {
  NetworkError(String message, {dynamic details})
      : super(message, code: 'NETWORK_ERROR', details: details);
}

class ValidationError extends AppError {
  ValidationError(String message, {dynamic details})
      : super(message, code: 'VALIDATION_ERROR', details: details);
}

File: apps/admin/src/lib/errors/app-error.ts
/**
 * Custom error classes for Admin app
 * Structured error handling with error codes
 */

export class AppError extends Error {
  constructor(
    message: string,
    public code: string,
    public statusCode: number = 500,
    public details?: unknown
  ) {
    super(message);
    this.name = 'AppError';
  }
}

export class AuthError extends AppError {
  constructor(message: string, details?: unknown) {
    super(message, 'AUTH_ERROR', 401, details);
    this.name = 'AuthError';
  }
}

export class PermissionError extends AppError {
  constructor(message: string, details?: unknown) {
    super(message, 'PERMISSION_ERROR', 403, details);
    this.name = 'PermissionError';
  }
}

export class ValidationError extends AppError {
  constructor(message: string, details?: unknown) {
    super(message, 'VALIDATION_ERROR', 400, details);
    this.name = 'ValidationError';
  }
}

export class NotFoundError extends AppError {
  constructor(message: string, details?: unknown) {
    super(message, 'NOT_FOUND', 404, details);
    this.name = 'NotFoundError';
  }
}

export function handleError(error: unknown): AppError {
  if (error instanceof AppError) {
    return error;
  }

  if (error instanceof Error) {
    return new AppError(error.message, 'UNKNOWN_ERROR', 500, error);
  }

  return new AppError('An unexpected error occurred', 'UNKNOWN_ERROR', 500, error);
}
